assets/app/scripts/services/api.js

```diff
@@ -0,0 +1,255 @@
+    }
+    Logger.warn('Invalid apiVersion "' + apiVersion + '"');
+    return undefined;
+  };
+  
+  var objectToResourceGroupVersion = function(apiObject) {
+    if (!apiObject || !apiObject.kind || !apiObject.apiVersion) {
+      return undefined;
+    }
+    var resource = kindToResource(apiObject.kind);
+    if (!resource) {
+      return undefined;
+    }
+    var groupVersion = parseGroupVersion(apiObject.apiVersion);
+    if (!groupVersion) {
+      return undefined;
+    }
+    return new ResourceGroupVersion(resource, groupVersion.group, groupVersion.version);
+  };
+  
+  // deriveTargetResource figures out the fully qualified destination to submit the object to.
+  // if resource is a string, and the object's kind matches the resource, the object's group/version are used.
+  // if resource is a ResourceGroupVersion, and the object's kind and group match, the object's version is used.
+  // otherwise, resource is used as-is.
+  var deriveTargetResource = function(resource, object) {
+    if (!resource || !object) {
+      return undefined;
+    }
+    var objectResource = kindToResource(object.kind);
+    var objectGroupVersion = parseGroupVersion(object.apiVersion);
+    var resourceGroupVersion = toResourceGroupVersion(resource);
+    if (!objectResource || !objectGroupVersion || !resourceGroupVersion) {
+      return undefined;
+    }
+    
+    // We specified something like "pods"
+    if (angular.isString(resource)) {
+      // If the object had a matching kind {"kind":"Pod","apiVersion":"v1"}, use the group/version from the object
+      if (resourceGroupVersion.equals(objectResource)) {
+        resourceGroupVersion.group = objectGroupVersion.group;
+        resourceGroupVersion.version = objectGroupVersion.version;
+      }
+      return resourceGroupVersion;
+    }
+    
+    // If the resource was already a fully specified object,
+    // require the group to match as well before taking the version from the object
+    if (resourceGroupVersion.equals(objectResource, objectGroupVersion.group)) {
+      resourceGroupVersion.version = objectGroupVersion.version;
+    }
+    return resourceGroupVersion;
+  };  
+  
+  // port of restmapper.go#kindToResource
+  var kindToResource = function(kind) {
+    if (!kind) {
+      return "";
+    }
+    var resource = String(kind).toLowerCase();
+    if (resource === 'endpoints' || resource === 'securitycontextconstraints') {
+      // no-op, plural is the singular
+    }
+    else if (resource[resource.length-1] === 's') {
+      resource = resource + 'es';
+    }
+    else if (resource[resource.length-1] === 'y') {
+      resource = resource.substring(0, resource.length-1) + 'ies';
+    }
+    else {
+      resource = resource + 's';
+    }
+
+    return resource;
+  };
+  
+  // apiInfo returns the host/port, prefix, group, and version for the given resource,
+  // or undefined if the specified resource/group/version is known not to exist.
+  var apiInfo = function(resource) {
```

Comment:
Worth considering using `return $q.when({ /*  */ })` so that when discovery lands we do not have to update calling code?